basic_module/staking/src/state.rs

```diff
@@ -0,0 +1,540 @@
+        }
+        Ok(())
+    }
+
+    pub fn sub_quantity(&mut self, delegatee: Public, quantity: StakeQuantity) -> Result<(), Error> {
+        if quantity != 0 {
+            if let Entry::Occupied(mut entry) = self.delegatees.entry(delegatee) {
+                let delegation = entry.get();
+                match delegation.cmp(&quantity) {
+                    Ordering::Greater => {
+                        *entry.get_mut() -= quantity;
+                        Ok(())
+                    }
+                    Ordering::Equal => {
+                        entry.remove();
+                        Ok(())
+                    }
+                    Ordering::Less => Err(Error::InsufficientStakes(Insufficient {
+                        required: quantity,
+                        actual: *delegation,
+                    })),
+                }
+            } else {
+                Err(Error::DelegateeNotFoundInCandidates(delegatee))
+            }
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn get_quantity(&self, delegatee: &Public) -> StakeQuantity {
+        self.delegatees.get(delegatee).cloned().unwrap_or(0)
+    }
+
+    pub fn into_iter(self) -> btree_map::IntoIter<Public, StakeQuantity> {
+        self.delegatees.into_iter()
+    }
+
+    pub fn sum(&self) -> u64 {
+        self.delegatees.values().sum()
+    }
+}
+
+pub struct Stakeholders(BTreeSet<Public>);
+
+impl Stakeholders {
+    pub fn load() -> Stakeholders {
+        Stakeholders(load_with_key(&*STAKEHOLDERS_KEY).unwrap_or_default())
+    }
+
+    pub fn save(self) {
+        let key = &*STAKEHOLDERS_KEY;
+        if !self.0.is_empty() {
+            write_with_key(key, self.0)
+        } else {
+            remove_key(key)
+        }
+    }
+
+    pub fn delegatees() -> HashMap<Public, StakeQuantity> {
+        Stakeholders::load().0.into_iter().fold(HashMap::new(), |mut map, stakeholder| {
+            let delegation = Delegation::load(&stakeholder);
+            delegation.into_iter().for_each(|(delegatee, quantity)| {
+                *map.entry(delegatee).or_default() += quantity;
+            });
+            map
+        })
+    }
+
+    pub fn update_by_increased_balance(&mut self, account: &StakeAccount) {
+        if account.balance > 0 {
+            self.0.insert(*account.public);
+        }
+    }
+
+    pub fn update_by_decreased_balance(&mut self, account: &StakeAccount, delegation: &Delegation) {
+        assert!(account.public == delegation.delegator);
+        if account.balance == 0 && delegation.sum() == 0 {
+            self.0.remove(account.public);
+        }
+    }
+}
+
+pub struct NextValidators(Vec<Validator>);
+impl NextValidators {
+    pub fn load() -> Self {
+        NextValidators(load_with_key(&*NEXT_VALIDATORS_KEY).unwrap_or_default())
+    }
+
+    pub fn save(self) {
+        write_with_key(&*NEXT_VALIDATORS_KEY, self.0)
+    }
+
+    pub fn elect() -> Self {
```

Comment:
FYI, currently, @majecty changes the way to store the validators' public keys.
https://github.com/CodeChain-io/foundry/pull/388

Reply:
I'll check it out and apply the change.